nonfree/nvidia-container-toolkit/nvidia-persistenced-wrapper/main.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"errors"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	stateFolder = "/var/run/nvidia-persistenced"
+	pidFile     = stateFolder + "/" + "nvidia-persistenced.pid"
+)
+
+func main() {
+	// first check if the pid file exists,
+	// then check if the process is running,
+	// if running try to kill it, then start the new process
+	if _, err := os.Stat(pidFile); err == nil {
+		// pid file exists, get the pid and kill the process
+		if pid, err := getProcessId(); err == nil {
+			if err := killProcess(pid); err != nil {
+				log.Printf("nvidia-persistenced-wrapper: error killing process: %d%v\n", pid, err)
+			}
+		} else {
+			log.Printf("nvidia-persistenced-wrapper: error reading pid file: %s%v\n", pidFile, err)
+			os.Exit(1)
+		}
+		// now we can remove the state directory
+		if err := os.RemoveAll(stateFolder); err != nil {
+			log.Printf("nvidia-persistenced-wrapper: error removing state directory: %s%v\n", stateFolder, err)
+			os.Exit(1)
+		}
+	} else {
+		// handle errors
+		if !errors.Is(err, os.ErrNotExist) {
+			log.Printf("nvidia-persistenced-wrapper: failed to stat pid file: %s%v\n", pidFile, err)
+			os.Exit(1)
+		}
+	}
+
+	cmd := exec.Command("/usr/local/bin/nvidia-persistenced",
+		[]string{
+			"--no-persistence-mode",
+			"--verbose",
+		}...)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		log.Printf("nvidia-persistenced-wrapper: error starting nvidia-persistenced: %v\n", err)
+		os.Exit(1)
+	}
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, unix.SIGINT, unix.SIGTERM)
+	if err := cmd.Process.Signal(<-ch); err != nil {
+		log.Printf("nvidia-persistenced-wrapper: error sending signal to nvidia-persistenced: %v\n", err)
+		os.Exit(1)
+	}
```

Comment:
we might need to wait for the child process to terminate here?

Reply:
good point, in my testing it has always successfuly terminated the nvidia process, I'd just add a check to be sure

Reply:
fixed